testsuite/MDAnalysisTests/analysis/test_lineardensity.py

```diff
@@ -43,138 +46,219 @@ def test_invalid_grouping():
+
+    # Assigning the Charges & Masses
+    u.add_TopologyAttr('charges', values=charges)
+    u.add_TopologyAttr('masses', values=masses)
+
+    u.trajectory = get_reader_for(coords)(coords,
+                                          order='fac',
+                                          n_atoms=n_atoms)
+
+    for ts in u.trajectory:
+        ts.dimensions = np.array([1, 1, 1, 90, 90, 90])
+
+    return u
+
+
+### Defining the Systems ###
+
+# The masses are all set to 1
+# The charges are dependent on each system
+# The positions are randomly taken from a uniform distribution between 0 and 1
+# NOTE: The positions at each time frame are independent on the previous time step
+
+def neutral_Particles(n_atoms, n_frames, atomsPerRes, resPerSegs):
+    charges = np.zeros(n_atoms)
+    masses = np.ones(n_atoms)
+
+    shape = (n_frames, n_atoms, 3)
+    coords = np.random.random(shape)
+
+    return make_Universe(coords, charges, masses, n_atoms, n_frames, atomsPerRes, resPerSegs)
+
+def charged_Particles(n_atoms, n_frames, atomsPerRes, resPerSegs):
+    charges = np.random.random(n_atoms) * 2 - np.ones(n_atoms) # Between -1 and 1
+    masses = np.ones(n_atoms)
+
+    shape = (n_frames, n_atoms, 3)
+    coords = np.random.random(shape)
+
+    return make_Universe(coords, charges, masses, n_atoms, n_frames, atomsPerRes, resPerSegs)
+
+def charged_Dimers(n_dimers, n_frames, dimersPerRes, resPerSegs, dimerLength = 0.05):
+    n_atoms = 2 * n_dimers
+    
+    charges = np.random.random(n_atoms) * 2 - np.ones(n_atoms) # Between -1 and 1
+    masses = np.ones(n_atoms)
+
+    shape = (n_frames, n_dimers, 3)
+    coords = np.random.random(shape) * 0.9 + np.ones(shape) * 0.05
+    # Puts in the same coordinate twice per dimer
+    coords = np.repeat(coords, 2, axis = 1)
+
+    # Shifts one of the atoms of each dimer by their bondLength in a random direction (defined to be in the box)
+    for time in coords:
+        for coord in time[::2,:]:
+            phi = np.random.random() * 2 * np.pi
+            theta = np.random.random() * np.pi
+            x = np.array([np.sin(theta)*np.cos(phi), np.sin(theta)*np.sin(phi), np.cos(theta)]) * dimerLength
+            coord += np.array([np.sin(theta)*np.cos(phi), np.sin(theta)*np.sin(phi), np.cos(theta)]) * dimerLength
 
-# test data for grouping='fragments'
-expected_masses_fragments = np.array(
-    [18.0154, 18.0154, 18.0154, 18.0154, 18.0154]
-)
-expected_charges_fragments = np.array([0, 0, 0, 0, 0])
-expected_xmass_fragments = np.array(
-    [0.0, 0.0, 0.0, 0.00717967, 0.00478644, 0.0, 0.0, 0.0, 0.0, 0.0]
-)
-expected_xcharge_fragments = np.array(
-    [0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0]
-)
 
+    return make_Universe(coords, charges, masses, n_atoms, n_frames, dimersPerRes * 2, resPerSegs)
 
+
+### Calculating the Expected Values ###
+
+def calc_Prop(u, prop = 'masses'): # Property can be 'masses' or 'charges'
+    expected_atoms = eval(f'u.atoms.{prop}')
```

Comment:
The [linter](https://github.com/MDAnalysis/mdanalysis/actions/runs/14314381402/job/40116958534#step:5:124) (and most programmers) would heavily discourage using `eval()`.  It could be used to run arbitrary code. Maybe you can use `getattr()` instead? (https://docs.python.org/3/library/functions.html#getattr)

```
getattr(u.atoms, prop)
```

There are quite a few `eval()` below you could replace similarly.